### PR TITLE
account creation through api sdk

### PR DESCRIPTION
**Pull request by tanishgupta1**

### CODE REVIEW

File: qds_sdk/commands.py [hunk 1]
```diff
@@ -59,7 +59,7 @@ def create(cls, **kwargs):
         """
         Create a command object by issuing a POST request to the /command endpoint
         Note - this does not wait for the command to complete
-
+{slave,type: arg}
```
Comment (mindprince) [inline qds_sdk/commands.py]: What is this?

File: qds_sdk/account.py [hunk 2]
```diff
@@ -3,8 +3,55 @@
...
+                return None
+        v = {}
+        options.level = "free"
+        options.compute_type = "CUSTOMER_MANAGED"
+        options.storage_type = "CUSTOMER_MANAGED"
+        options.CacheQuotaSizeInGB = "25"
+
+        v['account'] = vars(options)
+        return v
+
```
Comment (mindprince) [inline qds_sdk/account.py]: Clean up all the extra lines.

File: qds_sdk/account.py [hunk 3]
```diff
@@ -3,8 +3,55 @@
...
-
+from qds_sdk.qubole import Qubole
+from qds_sdk.exception import ParseError
+from qds_sdk.util import GentleOptionParser
+from qds_sdk.util import OptionParsingError
+from qds_sdk.util import OptionParsingExit
 
 class Account(SingletonResource):
+    usage = ("account <create> [options]")
+    optparser = GentleOptionParser(usage=usage)
```
Comment (mindprince) [inline qds_sdk/account.py]: We are not using optparser for any new code. Please use argparser. Please see the reports api for examples on how to add a new endpoint.

Comment (tanishgupta1) [inline qds_sdk/account.py]: fixed.

**Comment by mindprince** [comment]
Need to add tests.

File: bin/qds.py [hunk 4]
```diff
@@ -390,6 +393,29 @@ def clustermain(args):
...
+    arguments = accountclass._parse_account(args, action = "create")
+    if arguments is not None:
+        result = accountclass.create(**arguments)
+    return 0
+
+
+
+def accountmain(args):
+    accountclass = Account
+    actionset = set(["create"])
```
Comment (mindprince) [inline bin/qds.py]: Please don't do it like this. For all new endpoints we are following the approach used in `reportmain` or `schedulemain` etc. where all the parsing is left to Argparse.

File: qds_sdk/account.py [hunk 5]
```diff
@@ -1,10 +1,64 @@
 """
 The Accounts module contains the base definition for a Qubole account object
 """
-
+import json
 from qds_sdk.resource import SingletonResource
+from qds_sdk.qubole import Qubole
+from qds_sdk.exception import ParseError
+from qds_sdk.util import GentleOptionParser
```
Comment (mindprince) [inline qds_sdk/account.py]: Clean up unused imports.

File: tests/test_account.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+                                             'level': 'free', 'compute_type': 'CUSTOMER_MANAGED',
+                                             'aws_region': None, 'storage_type': 'CUSTOMER_MANAGED',
+                                             'CacheQuotaSizeInGB': '25', 'secret': None,
+                                             'use_previous_account_plan': None,
+                                             'compute_secret_key': None,
+                                             'compute_access_key': None, 'defloc': None}}
+                                            )
+
+
+    def test_account_create_invalid(self):
```
Comment (mindprince) [inline tests/test_account.py]: Need to add more test. One for each option like `--name` etc.

File: qds_sdk/account.py [hunk 7]
```diff
@@ -1,10 +1,64 @@
...
+        argparser = argparse.ArgumentParser(prog="qds.py account",
+                description="Account Creation for Qubole Data Service.")
+        subparsers = argparser.add_subparsers(title = "account operation")
+
+        #Create
+        create = subparsers.add_parser("create",
+                                       help="Create a new account")
+        create.add_argument("--name", dest="name", help="account name")
+        create.add_argument("--location", dest="defloc", help="Default location of S3")
+        create.add_argument("--access", dest = "acc_key", help="access key")
```
Comment (mindprince) [inline qds_sdk/account.py]: Need better help message, for example: `AWS Access Key ID`, `AWS Secret Access Key`.

File: qds_sdk/account.py [hunk 8]
```diff
@@ -1,10 +1,64 @@
...
+        create.add_argument("--secret", dest= "secret", help="secret key")
+        create.add_argument("--compute_access", dest="compute_access_key", help="compute access key")
+        create.add_argument("--compute_secret", dest="compute_secret_key", help="compute secret key")
+        create.add_argument('--aws', dest="aws_region", help="aws region")
```
Comment (mindprince) [inline qds_sdk/account.py]: This argument should be called `--region` or `--aws-region`.

File: qds_sdk/account.py [hunk 9]
```diff
@@ -1,10 +1,64 @@
...
+        subparsers = argparser.add_subparsers(title = "account operation")
+
+        #Create
+        create = subparsers.add_parser("create",
+                                       help="Create a new account")
+        create.add_argument("--name", dest="name", help="account name")
+        create.add_argument("--location", dest="defloc", help="Default location of S3")
+        create.add_argument("--access", dest = "acc_key", help="access key")
+        create.add_argument("--secret", dest= "secret", help="secret key")
+        create.add_argument("--compute_access", dest="compute_access_key", help="compute access key")
```
Comment (mindprince) [inline qds_sdk/account.py]: Need to highlight the difference between compute access key and access key.

Also, don't use underscore in the `--` arguments. Use hypens.

File: qds_sdk/account.py [hunk 10]
```diff
@@ -1,10 +1,64 @@
...
+        create.add_argument("--compute_secret", dest="compute_secret_key", help="compute secret key")
+        create.add_argument('--aws', dest="aws_region", help="aws region")
+        create.add_argument("--previous_account_plan", dest="use_previous_account_plan",choices=["true", "false"], help="either true or false")
```
Comment (mindprince) [inline qds_sdk/account.py]: Better help message here.

File: qds_sdk/account.py (hunk 7 above)
Comment (mindprince) [inline qds_sdk/account.py]: Also, don't call it just `--access`, it needs to be more descriptive.

File: qds_sdk/account.py [hunk 11]
```diff
@@ -1,10 +1,64 @@
...
+        parser = AccountCmdLine.parsers()
+        parsed = parser.parse_args(args)
+        return parsed.func(parsed)
+
+    @staticmethod
+    def create(args):
+        args.level = "free"
+        args.compute_type = "CUSTOMER_MANAGED"
+        args.storage_type = "CUSTOMER_MANAGED"
+        args.CacheQuotaSizeInGB = "25"
```
Comment (mindprince) [inline qds_sdk/account.py]: Shouldn't there be an option for this?

File: qds_sdk/account.py [hunk 12]
```diff
@@ -1,10 +1,64 @@
...
+
+    @staticmethod
+    def run(args):
+        parser = AccountCmdLine.parsers()
+        parsed = parser.parse_args(args)
+        return parsed.func(parsed)
+
+    @staticmethod
+    def create(args):
+        args.level = "free"
```
Comment (mindprince) [inline qds_sdk/account.py]: What is this?

**Comment by mindprince** [comment]
cc - @itsvikramagr Please review.

File: qds_sdk/account.py (hunk 12 above)
Comment (tanishgupta1) [inline qds_sdk/account.py]: I have only added those options which are present in UI while adding new account and there are some other params which are passed too but are not enter by user like level , compute_type , storage_type and CacheQuotaSizeInGB. For these options default values are already set in js file so I added same values here too.

File: tests/test_account.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                                             'compute_access_key': None, 'defloc': None}}
+                                            )
+
+    def test_aws_region(self):
+        sys.argv = ['qds.py', 'account', 'create','--aws-region', 'us-east-1']
+        print_command()
+        Connection._api_call = Mock(return_value={})
+        qds.main()
+        Connection._api_call.assert_called_with("POST", "account", {'account':
+                                            {'acc_key': None,
```
Comment (mindprince) [inline tests/test_account.py]: Please follow PEP8 formatting/indentation style.

There are some existing files which are not following PEP8 completely. But we are trying to follow it as closely as we can in the new files which we are adding. So, please run `pep8` on both account.py and test_account.py.

Also, consider using an IDE like [pycharm](https://www.jetbrains.com/pycharm/) which suggest these improvements when you are typing.

File: tests/test_account.py [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+import qds
+from qds_sdk.connection import Connection
+from test_base import print_command
+from test_base import QdsCliTestCase
+
+
+class TestAccountCreate(QdsCliTestCase):
+
+    def test_none(self):
+        sys.argv = ['qds.py', 'account', 'create']
```
Comment (mindprince) [inline tests/test_account.py]: When I tried this:
```
qds.py --vv account create
```
I got
```
{"error":"Error creating user"}
Error: Status code 422 (ResourceInvalid) from url https://api.qubole.com/api/v1.2/account
```

File: qds_sdk/account.py [hunk 15]
```diff
@@ -1,10 +1,59 @@
...
+    def parsers():
+        argparser = argparse.ArgumentParser(prog="qds.py account",
+                                            description="Account Creation for Qubole Data Service.")
+        subparsers = argparser.add_subparsers(title = "account operation")
+
+        #Create
+        create = subparsers.add_parser("create",
+                                       help="Create a new account")
+        create.add_argument("--name", dest="name", help="Account name")
+        create.add_argument("--location", dest="defloc", help="Default location of S3")
```
Comment (mindprince) [inline qds_sdk/account.py]: Which of these are required arguments?
We should add required=true for those.

**Comment by mindprince** [comment]
Please also rebase this on the current master branch.

File: qds_sdk/account.py [hunk 16]
```diff
@@ -1,10 +1,59 @@
...
+        create = subparsers.add_parser("create",
+                                       help="Create a new account")
+        create.add_argument("--name", dest="name", required=True, help="Account name")
+        create.add_argument("--location", dest="defloc", required=True, help="Default location of S3")
+        create.add_argument("--storage-access-key", dest = "acc_key", required=True, help="AWS Access Key for storage ")
+        create.add_argument("--storage-secret-key", dest= "secret", required=True, help="AWS Secret Key for storage")
+        create.add_argument("--compute-access-key", dest="compute_access_key", required=True, help="AWS Access Key for compute")
+        create.add_argument("--compute-secret-key", dest="compute_secret_key", required=True, help="AWS Secret Key for compute")
+        create.add_argument('--aws-region', dest="aws_region", required=True, help="AWS Region")
+        create.add_argument("--previous-account-plan", dest="use_previous_account_plan",choices=["true", "false"], required=True, help="Use previous account plan, either true or false")
```
Comment (mindprince) [inline qds_sdk/account.py]: Why is this a required option?

File: tests/test_account.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+import qds
+from qds_sdk.connection import Connection
+from test_base import print_command
+from test_base import QdsCliTestCase
+
+
+class TestAccountCreate(QdsCliTestCase):
+
+    def test_all(self):
+        sys.argv = ['qds.py', 'account', 'create', '--name', 'new_account', '--location', 's3://dev.canopydata.com' ,
```
Comment (mindprince) [inline tests/test_account.py]: Don't use internal S3 bucket names.

**Comment by mindprince** [comment]
Squash merged the PR after making some edits. Commit: https://github.com/qubole/qds-sdk-py/commit/49345cc1bb1a6c41a1002b29d2c3f61c53da06d9

**Pull request closed without merging**